docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+
+1. Enqueue in memory
+2. Enqueue persistently by writing out a deletion list, storing the
+   attachment generations and node generation.
+3. Validate the deletion list by calling to the control plane
+4. Execute the valid parts of the deletion list (i.e. call DeleteObjects)
+
+Note that steps 2 and 3 can be swapped: it depends on whether we would like
+to optimize for delaying validation for a long time in bounded memory (persist
+first, delete later), or optimize for I/O (validate first, then persist).
+
+There is existing work (https://github.com/neondatabase/neon/pull/4960) to
+create a deletion queue: this would be extended by adding the "step 3" validation
+step.
+
+Since its reason for existence is optimization rather than correctness,
+there is a lot of flexibility in exactly how the deletion queue should work,
+as long as it obeys the rule to validate generations before executing deletions:
+
+- Option A (simplest):validate the
+  deletion list before persisting it. This has the downside of delaying
+  persistence of deletes until the control plane is available.
+- Option B: validate a list at the point of executing it. This has the downside
+  that because we're doing the validation much later, there is a good chance
+  some attachments might have changed, and we will end up leaking objects.
+- Option C (preferred): validate lazily in the background after persistence of the list, maintaining
+  an "executable" pointer into the list of deletion lists to reflect
+  which ones are elegible for execution, and re-writing deletion lists if
+  they are found to contain some operations not elegible for execution. This
+  is the most efficient approach in terms of I/O, at the cost of some complexity
+  in implementation.
+
+As well as reducing leakage, the ability of a persistent queue to accumulate
+deletions over long timescale has side benefits:
+
+- Over enough time we will always accumulate enough keys to issue full-sized
+  (1000 object) DeleteObjects requests, minimizing overall request count.
+- If in future we implement a read-only mode for pageservers to read the
+  data of tenants, then delaying deletions avoids the need for that remote
+  pageserver to be fully up to date with the latest index when serving reads
+  from old LSNs. The read-only node might be using stale metadata that refers
+  to objects eliminated in a recent compaction, but can still service reads
+  without refreshing metadata as long as the old objects deletion is delayed.
+
+#### Deletion queue persistence format
+
+_Note: the following format describes a persistent format for "Option C" in
+the previous section. It would be simpler in "Option A", as one would not
+need to store generation information for pre-validated deletes_
+
+Persistence is to S3 rather than local disk so that if a pageserver
+is stopped uncleanly and then the same node_id is started on a fresh
+physical machine, the deletion queue can still continue without leaking
+objects.
```

Comment:
How do we deal with split brain here, i.e., two processes running simultaneously with same node id?

Reply:
Revised this to use local disk for deletion queue, since node generation is removed, & thereby avoid any concerns about conflicts over remote objects.